File: src/uexchanges/forms/fingerprint.py
from __future__ import annotations

import hashlib
import json
from urllib.parse import urlsplit, urlunsplit

from .models import FormField


def canonicalize_form_url(url: str) -> str:
    """Remove fragments while preserving provider-significant query parameters."""
    parsed = urlsplit(url)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise ValueError("form URL must be absolute HTTP(S)")
    return urlunsplit((parsed.scheme.lower(), parsed.netloc.lower(), parsed.path or "/", parsed.query, ""))


def form_schema_fingerprint(*, provider: str, canonical_form_url: str, fields: tuple[FormField, ...]) -> str:
    """Hash structural form state only; answers/evidence never affect form identity."""
    if not provider.strip():
        raise ValueError("provider must be non-empty")
    payload = {
        "provider": provider.strip().lower(),
        "url": canonicalize_form_url(canonical_form_url),
        "fields": [
            {
                "field_key": field.field_key,
                "label": field.label,
                "field_type": field.field_type.value,
                "required": field.required,
                "options": list(field.options),
                "maxlength": field.maxlength,
            }
            for field in fields
        ],
    }
    encoded = json.dumps(payload, ensure_ascii=False, sort_keys=True, separators=(",", ":")).encode("utf-8")
    return f"sha256:{hashlib.sha256(encoded).hexdigest()}"

File: tools/form-executor/src/fingerprint.mjs
import crypto from 'node:crypto';

export function canonicalizeFormUrl(value) {
  if (typeof value !== 'string') throw new Error('form URL must be a string');
  const match = value.match(/^(https?):\/\/([^/?#]+)([^?#]*)(\?[^#]*)?(?:#.*)?$/i);
  if (!match) throw new Error('form URL must be absolute HTTP(S)');
  const scheme = match[1].toLowerCase();
  const netloc = match[2].toLowerCase();
  const path = match[3] || '/';
  const query = match[4] || '';
  return `${scheme}://${netloc}${path}${query}`;
}

function stableJson(value) {
  if (Array.isArray(value)) return `[${value.map(stableJson).join(',')}]`;
  if (value && typeof value === 'object') {
    const keys = Object.keys(value).sort();
    return `{${keys.map((key) => `${JSON.stringify(key)}:${stableJson(value[key])}`).join(',')}}`;
  }
  return JSON.stringify(value);
}

export function formSchemaFingerprint({ provider, canonicalFormUrl, fields }) {
  if (typeof provider !== 'string' || !provider.trim()) throw new Error('provider must be non-empty');
  if (!Array.isArray(fields)) throw new Error('fields must be an array');
  const payload = {
    provider: provider.trim().toLowerCase(),
    url: canonicalizeFormUrl(canonicalFormUrl),
    fields: fields.map((field) => ({
      field_key: field.field_key,
      label: field.label,
      field_type: field.field_type,
      required: Boolean(field.required),
      options: Array.isArray(field.options) ? field.options : [],
      maxlength: field.maxlength ?? null,
    })),
  };
  return `sha256:${crypto.createHash('sha256').update(stableJson(payload), 'utf8').digest('hex')}`;
}
